### ISSUE TITLE
`terraform test` expects variables from repo root when `module` reference is specified

### ISSUE THREAD

**Issue by drewmullen** [hashicorp/terraform#33626]
### Terraform Version

```shell
./tftest -v
Terraform v1.6.0-alpha20230802
on darwin_arm64
+ provider registry.terraform.io/hashicorp/aws v5.9.0
+ provider registry.terraform.io/hashicorp/awscc v0.56.0
```

### Terraform Configuration Files

See PR here: https://github.com/aws-ia/terraform-aws-vpc/pull/127/files

### Debug Output

n/a

### Expected Behavior

I am trying to setup tftest files that call my examples directory root modules. This feeds 3 birds with 1 scone:
- write tests that test known / expected use cases for the module
- validate the examples we provide users are functional
- keep code DRY

```hcl
run "validate_public_private_example_root" {
  command = plan

  module {
    source = "./examples/public_private_flow_logs"
  }
  ...
}
```

^ this should call my example directory `public_private_flow_logs` root and plan that out. Once `apply` is available I would build that example and then validate certain outcomes

### Actual Behavior

terraform test seemingly expects inputs for variables in the root dir which is strange because my module call defines those values. Its like its trying to both instantiate the module reference in my tftest.hcl file (points to an example dir) and another instantiation of the root of my module.

Output:
```
$ ./tftest test
tests/test.tftest.hcl... fail
╷
│ Warning: Argument is deprecated
│
│   with aws_eip.nat,
│   on main.tf line 99, in resource "aws_eip" "nat":
│   99:   vpc      = true
│
│ use domain attribute instead
╵
╷
│ Error: No value for required variable
│
│   on variables.tf line 1:
│    1: variable "name" {
│
│ The root module input variable "name" is not set, and has no default value. Use a -var or -var-file command line
│ argument to provide a value for this variable.
╵
╷
│ Error: No value for required variable
│
│   on variables.tf line 18:
│   18: variable "az_count" {
│
│ The root module input variable "az_count" is not set, and has no default value. Use a -var or -var-file command line
│ argument to provide a value for this variable.
╵
╷
│ Error: No value for required variable
│
│   on variables.tf line 105:
│  105: variable "subnets" {
│
│ The root module input variable "subnets" is not set, and has no default value. Use a -var or -var-file command line
│ argument to provide a value for this variable.
╵
  run "validate_public_private_example_root"... skip

Failure! 0 passed, 0 failed, 1 skipped.
```

Notice all the vars mentioned above are defined in the example dir root module call: https://github.com/aws-ia/terraform-aws-vpc/blob/main/examples/public_private_flow_logs/main.tf#L9,7,8,11

### Steps to Reproduce

```shell
$ git clone git@github.com:drewmullen/terraform-aws-vpc-1.git
$ cd terraform-aws-vpc-1
$ git checkout terraform-test-native
$ terraform init
$ terraform test 
```

### Additional Context

LMK if this is not an intended use case for `test`. I believe writing tests that validate your example/ dirs will be the first major success for module developers. Once those are implemented we can add more for unit testing locals, etc

**Comment by liamcervante** [comment]
I've changed this from a bug as the behaviour is expected, but will leave it open as an enhancement and we can see about improving the messaging here to make it clearer why the variables are required.

**Comment by drewmullen** [comment]
Hi @liamcervant. Thanks for the thorough explanation. I imagine it will be useful to more than just me in the future! I also appreciate you evaluating the use case and your positive feedback.

Alright so I understand the root cause and that this is intended. Silly question next... how do I set variables for a module I'm not calling? In y'alls tests I found [this example](https://github.com/hashicorp/terraform/blob/5fa956e1cb6a43e8971d0e867701f2686751158e/internal/command/testdata/test/variables/set_variables.tftest.hcl#L2C1-L4C4) and have attempted to replicate in my [latest commit](https://github.com/aws-ia/terraform-aws-vpc/pull/127/commits/81c51431e7814772c355efa546d6719465ff1104) but I'm still getting the same error.

Lastly, is `command = apply` supported yet? I saw in [this comment]( https://discuss.hashicorp.com/t/request-for-testing-terraform-test/56254/3?u=dmullen) that only `plan` was supported atm.

**Comment by drewmullen** [comment]
I had one more idea... perhaps the example dir needs to pass the defaults... I tried setting default vars in the example but am still getting the same error. Heres the [commit](https://github.com/aws-ia/terraform-aws-vpc/pull/127/commits/c511d5ba23629bea9dbb0f23fff0ba699f3c98ca)

**Comment by liamcervante** [comment]
> Lastly, is command = apply supported yet? I saw in [this comment](https://discuss.hashicorp.com/t/request-for-testing-terraform-test/56254/3?u=dmullen) that only plan was supported atm.

Yes, `command = apply` is supported now. It's actually the default, so if you don't specify the `command` attribute then you'll get an apply operation.

**Comment by drewmullen** [comment]
Alright so basic executions I have working now (huzzah!)

However, i somehow stumbled upon a weird issue that does seem to be related to the same context. Pls let me know if you want this in a new issue. I added a 2nd `run {}` that actually applies the config. everything seemed to work great once setting the `variables {}` block. ✅ 

However, after adding a `depends_on` to my module reference and then all hell seemed to break loose. [commit](https://github.com/aws-ia/terraform-aws-vpc/pull/127/commits/9b391fef80f0c72f9c712f27b971c7b3338a2c21)

```shell
$ ./tftest test
tests/test.tftest.hcl... fail

│ Error: Invalid for_each argument
│ 
│   on main.tf line 58, in resource "aws_subnet" "public":
│   58:   for_each = contains(local.subnet_keys, "public") ? toset(local.azs) : toset([])
│     ├────────────────
│     │ local.azs is a list of string, known only after apply
│     │ local.subnet_keys is tuple with 2 elements
│ 
│ The "for_each" set includes values derived from resource attributes that cannot be determined until apply, and so Terraform cannot determine the full set of keys that will identify the instances of this resource.
│ 
│ When working with unknown values in for_each, it's better to use a map value where the keys are defined statically in your configuration and where only the values contain apply-time results.
│ 
│ Alternatively, you could use the -target planning option to first apply only the resources that the for_each value depends on, and then apply a second time to fully converge.
╵
╷
│ Error: Invalid for_each argument
│ 
│   on main.tf line 78, in resource "aws_route_table" "public":
│   78:   for_each = contains(local.subnet_keys, "public") ? toset(local.azs) : toset([])
│     ├────────────────
│     │ local.azs is a list of string, known only after apply
│     │ local.subnet_keys is tuple with 2 elements
│ 
│ The "for_each" set includes values derived from resource attributes that cannot be determined until apply, and so Terraform cannot determine the full set of keys that will identify the instances of this resource.
│ 
│ When working with unknown values in for_each, it's better to use a map value where the keys are defined statically in your configuration and where only the values contain apply-time results.
│ 
│ Alternatively, you could use the -target planning option to first apply only the resources that the for_each value depends on, and then apply a second time to fully converge.
╵
╷
│ Error: Invalid for_each argument
│ 
│   on main.tf line 98, in resource "aws_eip" "nat":
│   98:   for_each = toset(local.nat_configuration)
│     ├────────────────
│     │ local.nat_configuration is a list of string, known only after apply
│ 
│ The "for_each" set includes values derived from resource attributes that cannot be determined until apply, and so Terraform cannot determine the full set of keys that will identify the instances of this resource.
│ 
│ When working with unknown values in for_each, it's better to use a map value where the keys are defined statically in your configuration and where only the values contain apply-time results.
│ 
│ Alternatively, you could use the -target planning option to first apply only the resources that the for_each value depends on, and then apply a second time to fully converge.
╵
╷
│ Error: Invalid for_each argument
│ 
│   on main.tf line 232, in resource "aws_subnet" "private":
│  232:   for_each = toset(try(local.private_per_az, []))
│     ├────────────────
│     │ local.private_per_az will be known only after apply
│ 
│ The "for_each" set includes values derived from resource attributes that cannot be determined until apply, and so Terraform cannot determine the full set of keys that will identify the instances of this resource.
│ 
│ When working with unknown values in for_each, it's better to use a map value where the keys are defined statically in your configuration and where only the values contain apply-time results.
│ 
│ Alternatively, you could use the -target planning option to first apply only the resources that the for_each value depends on, and then apply a second time to fully converge.
╵
╷
│ Error: Invalid for_each argument
│ 
│   on main.tf line 256, in resource "aws_route_table" "private":
│  256:   for_each = toset(try(local.private_per_az, []))
│     ├────────────────
│     │ local.private_per_az will be known only after apply
│ 
│ The "for_each" set includes values derived from resource attributes that cannot be determined until apply, and so Terraform cannot determine the full set of keys that will identify the instances of this resource.
│ 
│ When working with unknown values in for_each, it's better to use a map value where the keys are defined statically in your configuration and where only the values contain apply-time results.
│ 
│ Alternatively, you could use the -target planning option to first apply only the resources that the for_each value depends on, and then apply a second time to fully converge.
╵
  run "apply_validate_public_private_example_root"... skip

Failure! 0 passed, 1 failed, 1 skipped.
```

if i [comment out the `depends_on`](https://github.com/aws-ia/terraform-aws-vpc/pull/127/files#diff-d3ca3fc6c4a9eb427fc55a2ce76e02b43bc6498495b75b59c9099b13cbe009ffR42) the exact same code base works fine

**Comment by liamcervante** [comment]
I think that issue isn't related to `terraform test` specifically. I checked out your changes, and if I try and plan the example configuration outside of the test context (ie. just with a normal `terraform plan` command) I get the same set of errors.

Any variables you use within a `foreach` or `count` block have to be known at plan time in order for the plan to actually be able to report the proposed changes properly. The `depends_on` attribute complicates this because it forces Terraform to anticipate that it should only calculate values after the depends_on thing is created, which means all the values in the module are unknown at plan time and you get the cascade of errors.

If you were applying your example for real you could work around this by running two commands in the following order:

```
terraform apply -target=aws_ssm_parameter.foo
terraform apply
```

The first command will make sure the resource that everything depends on is created, then the follow up apply won't have the problems it did originally, because the resource it depends on already exists so the plan happens like normal.

Normally, you would be able to do the same in your test file with the `plan_options` block:

```hcl
run "setup_depends_on" {
  plan_options {
    target = [
      aws_ssm_parameter.foo,
    ]
  }
}

run "plan_validate_public_private_example_root" {
 // ...
}
```

However, in your case this won't work because you're loading the example configuration as an external module (via the `module` block). This means the state from each `run` block isn't being copied between run blocks. There is reasoning behind this, which is quite long and has a history, but the short version is that we didn't see a use case for when you would want to maintain the state between run blocks for alternate modules being loaded. With that in mind we thought it simpler to just have alternate modules keep their state private, while any run blocks that execute against the main configuration could share their state. Trying to maintain and propagate multiple states is difficult, because if run blocks execute against the state for a different module you'd get weird behaviour (like resources being deleted and created unexpectedly).

In the short term, you could move your tests for the example into `examples/public_private_flow_logs/tests` and execute them from the example configuration directly. This will allow you to remove your `module` blocks and then the workaround with the `target` parameter I mentioned above should start working for you.

In the long term, I will revisit our decision on state management and see if we can make this use case work easily. We even have a callout for this in our draft docs so it's nice we got a use case so early in the alpha period (thanks!). This would enable you to move your tests all back into the same directory at some point in the future.

I hope this all makes sense! And thanks again for your contributions and feedback!

**Comment by drewmullen** [comment]
Perfectly clear. The commit with the param was me just trying to break stuff lol Appreciate the explanation.

Cool. I'll close this now and see about actually writing tests since the base functions seem to be working!

Ty!

**Comment by liamcervante** [comment]
Just a quick note here that we merged #33643 today, which should address some of the feedback in this issue.

From the next release:
  - You won't have to specify all the variables for the main configuration at the file level.
  - Run blocks that reference the same alternate modules will share state between them, so you will be able to execute a series of run blocks against an alternate module in the way this ticket expects.

**Comment by drewmullen** [comment]
thats awesome, @liamcervante. thank you for wrestling with the ideas with me!
